midend/replaceSelectRange.cpp

```diff
@@ -0,0 +1,159 @@
+        auto valConst = new IR::Constant(constType, min, base, true);
+        big_int mask = ~(range_size - 1) & ((((big_int) 1) << width) - 1);
+        auto maskConst = new IR::Constant(constType, mask, base, true);
+        auto m = new IR::Mask(r->srcInfo, valConst, maskConst);
+
+        masks.push_back(m);
+        range_size_remaining -= range_size;
+        min += range_size;
+    }
+
+    return masks;
+}
+
+std::vector<IR::Vector<IR::Expression>>
+DoReplaceSelectRange::cartesianAppend(const std::vector<IR::Vector<IR::Expression>>& vecs,
+                                      const std::vector<const IR::Mask *>& masks) {
+    std::vector<IR::Vector<IR::Expression>> newVecs;
+
+    for (auto v : vecs) {
+        for (auto mask : masks) {
+            auto copy(v);
+
+            copy.push_back(mask);
+            newVecs.push_back(copy);
+        }
+    }
+
+    return newVecs;
+}
+
+std::vector<IR::Vector<IR::Expression>>
+DoReplaceSelectRange::cartesianAppend(const std::vector<IR::Vector<IR::Expression>>& vecs,
+                                      const IR::Expression *e) {
+    std::vector<IR::Vector<IR::Expression>> newVecs;
+
+    for (auto v : vecs) {
+        auto copy(v);
+
+        copy.push_back(e);
+        newVecs.push_back(copy);
+    }
+
+    return newVecs;
+}
+
+const IR::Node* DoReplaceSelectRange::postorder(IR::SelectCase* sc) {
+    auto newCases = new IR::Vector<IR::SelectCase>();
+    auto keySet = sc->keyset;
+
+    if (auto r = keySet->to<IR::Range>()) {
+        auto masks = rangeToMasks(r);
+        if (masks.empty())
+            return sc;
+
+        for (auto mask : masks) {
+            auto c = new IR::SelectCase(sc->srcInfo, mask, sc->state);
+            newCases->push_back(c);
+        }
+
+        return newCases;
+    } else if (auto oldList = keySet->to<IR::ListExpression>()) {
+        std::vector<IR::Vector<IR::Expression>> newVectors;
+        IR::Vector<IR::Expression> first;
+
+        newVectors.push_back(first);
+
+        for (auto key : oldList->components) {
+            if (auto r = key->to<IR::Range>()) {
+                auto masks = rangeToMasks(r);
+                if (masks.empty())
+                    return sc;
+
+                newVectors = cartesianAppend(newVectors, masks);
+            } else {
+                newVectors = cartesianAppend(newVectors, key);
+            }
+        }
+
+        for (auto v : newVectors) {
+            auto le = new IR::ListExpression(oldList->srcInfo, v);
+            newCases->push_back(new IR::SelectCase(sc->srcInfo, le, sc->state));
+        }
+        if (newCases->size() > MAX_CASES) {
+            ::warning(ErrorType::ERR_OVERLIMIT,
+                      "select key set expression with a range expands into %2% "
+                      "ternary key set expressions, which may lead to run-time "
+                      "performance or parser configuration space issues in some"
+                      " targets.", sc, MAX_CASES);
```

Comment:
I would recommend replacing MAX_CASES in this line with newCases->size(), so that the warning message shows the actual number of ternary key set expressions expanded into.  As it is now, it will always print 100, even if the code causes an expansion into more than 100.